### PR TITLE
[15]: Implemented delete filter flow

### PR DESCRIPTION
**Pull request by IvanShetkoGodel**
Within this task were added:
-remove filter method in filter service class and controller's action
-unit tests for BL layer

### CODE REVIEW

File: GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL.Tests/Services/FilterServiceTests.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            Filter toRemove = null;
+
+            unitOfWork.Setup(x => x.Filters.GetById(id)).Returns(toRemove);
+
+            underTest.RemoveFilter(id);
+
+            unitOfWork.Verify(x => x.Filters.Delete(toRemove), Times.Never());
+            unitOfWork.Verify(x => x.SaveChanges(), Times.Never());
+        }
+
```
Comment (vladkuzmich) [inline GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL.Tests/Services/FilterServiceTests.cs]: @IvanShetkoGodel Why your test methods without comments?

File: GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL.Tests/Services/FilterServiceTests.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            Filter toRemove = new Filter() { Id = 1};
+
+            unitOfWork.Setup(x => x.Filters.GetById(id)).Returns(toRemove);
+
+            underTest.RemoveFilter(id);
+
+            unitOfWork.Verify(x => x.Filters.Delete(toRemove));
+            unitOfWork.Verify(x => x.SaveChanges());
+        }
+
```
Comment (vladkuzmich) [inline GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL.Tests/Services/FilterServiceTests.cs]: @IvanShetkoGodel The same. Why your test methods without comments?

File: GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL.Tests/Services/FilterServiceTests.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            Filter toRemove = new Filter() { Id = 1 };
+
+            unitOfWork.Setup(x => x.Filters.GetById(id)).Throws(new Exception()); 
+
+            underTest.RemoveFilter(id);
+
+            unitOfWork.Verify(x => x.RollBack());
+        }
+    }
+}
```
Comment (vladkuzmich) [inline GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL.Tests/Services/FilterServiceTests.cs]: @IvanShetkoGodel The same. Why your test methods without comments?

File: GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL/Services/FilterService.cs [hunk 4]
```diff
@@ -1,12 +1,34 @@
...
+                    unitOfWork.SaveChanges();
+                }
+            }
+            catch(Exception e)
+            {
+                logger.Error($"Remove filter {id} exception: {e.Message}");
+                unitOfWork.RollBack();
+            }
+        }
     }
```
Comment (vladkuzmich) [inline GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL/Services/FilterService.cs]: Rename your "toDelete" variable as "filter" or something like that.

File: GodelMastery.FleaMarket/GodelMastery.FleaMarket.Web/Controllers/FilterController.cs [hunk 5]
```diff
@@ -1,15 +1,28 @@
...
 using GodelMastery.FleaMarket.BL.Interfaces;
+using GodelMastery.FleaMarket.DAL.Interfaces;
 
 namespace GodelMastery.FleaMarket.Web.Controllers
 {
+    [Authorize]
     public class FilterController : Controller
     {
         private readonly IFilterService filterService;
+        private readonly IUnitOfWork unitOfWork;
```
Comment (vladkuzmich) [inline GodelMastery.FleaMarket/GodelMastery.FleaMarket.Web/Controllers/FilterController.cs]: Do you really need "unitOfWork" in controller?!

File: GodelMastery.FleaMarket/GodelMastery.FleaMarket.Web/Controllers/FilterController.cs [hunk 6]
```diff
@@ -1,15 +1,28 @@
...
+            this.unitOfWork = unitOfWork;
+        }
+        [HttpPost]
+        [ValidateAntiForgeryToken]
+        public ActionResult RemoveFilter(int id)
+        {
+            filterService.RemoveFilter(id);
+            return View("Home", "Dashboard");
         }
     }
```
Comment (vladkuzmich) [inline GodelMastery.FleaMarket/GodelMastery.FleaMarket.Web/Controllers/FilterController.cs]: @IvanShetkoGodel  Why your action has only post method?  Where is the confirmation letter for the user?

**Comment by vladkuzmich** [comment]
@IvanShetkoGodel view comments

File: GodelMastery.FleaMarket/GodelMastery.FleaMarket.Web/Controllers/FilterController.cs (hunk 6 above)
Comment (IvanShetkoGodel) [reply GodelMastery.FleaMarket/GodelMastery.FleaMarket.Web/Controllers/FilterController.cs]: @vladkuzmich Confirmation must be on the front end part, isn't it?

File: GodelMastery.FleaMarket/GodelMastery.FleaMarket.Web/Content/Error.css [hunk 7]
```diff
@@ -1,8 +1,12 @@
-﻿.error-page {
+﻿body {
     background: url("../Images/error.jpg") no-repeat center;
```
Comment (Zhunusau) [inline GodelMastery.FleaMarket/GodelMastery.FleaMarket.Web/Content/Error.css]: Remove this picture please, make normal error page, example https://cdn.discordapp.com/attachments/454633341073883136/462166277625085954/unknown.png

File: GodelMastery.FleaMarket/GodelMastery.FleaMarket.Web/Controllers/FilterController.cs [hunk 8]
```diff
@@ -23,5 +29,26 @@ public FilterController(IDashboardService dashboardService, IFilterViewModelFact
...
         }
+
+        [HttpGet]
+        public ActionResult RemoveFilter(int filterId)
+        {
+            var filter = filterService.GetFilterById(filterId);
+
+            if(filter != null )
+            {
+                return PartialView(filter);
```
Comment (Zhunusau) [inline GodelMastery.FleaMarket/GodelMastery.FleaMarket.Web/Controllers/FilterController.cs]: Use filterViewModel instead FilterDto.

File: GodelMastery.FleaMarket/GodelMastery.FleaMarket.Web/Views/Error/HandleExceptions.cshtml [hunk 9]
```diff
@@ -1,7 +1,7 @@
 ﻿@Styles.Render("~/Content/Error");
-<body class="error-page">
-    <div class="col-md-4 col-md-offset-6 text-danger">
+<div class="container pull-right">
+    <div class="text-danger">
         <h1>We will back soon</h1>
-        <h2 class="error-page-font">@ViewBag.Message</h2>
+        <h2 class="text-danger">@ViewBag.Message</h2>
```
Comment (Zhunusau) [inline GodelMastery.FleaMarket/GodelMastery.FleaMarket.Web/Views/Error/HandleExceptions.cshtml]: Make the error text on the page center.

File: GodelMastery.FleaMarket/GodelMastery.FleaMarket.Web/Views/Filter/RemoveFilter.cshtml [hunk 10]
```diff
@@ -0,0 +1,25 @@
+﻿@model GodelMastery.FleaMarket.BL.Dtos.FilterDto
+
+<div>
+    @using (Html.BeginForm())
```
Comment (Zhunusau) [inline GodelMastery.FleaMarket/GodelMastery.FleaMarket.Web/Views/Filter/RemoveFilter.cshtml]: Make the modal window on the page center, deleted cross "x"or make it in the right corner, and increase text.
image

**Comment by Zhunusau** [comment]
When you solve conflicts, use the helper CurrentUser.GetUserId to get the user id.

**Comment by vyshkouski-godeltech-intelliflo** [comment]
Make rebase on master please

File: GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL.Tests/Services/FilterServiceTests.cs [hunk 11]
```diff
@@ -190,5 +190,73 @@ public void Create_When_Filter_Added_With_Exception_Then_Operation_Throw_Excepti
...
+        public void RemoveFilter_When_Filter_Not_Exist_Should_Not_Call_Delete_And_Save_Methods()
+        {
+            //arrange
+            int id = 1;
+            Filter toRemove = null;
+
+
+            unitOfWork.Setup(x => x.Filters.GetById(id)).Returns(toRemove);
+
+            //when
```
Comment (vyshkouski-godeltech-intelliflo) [inline GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL.Tests/Services/FilterServiceTests.cs]: @IvanShetkoGodel Use 'act' please. Ask @vladkuzmich for more details. We've already discussed it today

File: GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL.Tests/Services/FilterServiceTests.cs [hunk 12]
```diff
@@ -190,5 +190,73 @@ public void Create_When_Filter_Added_With_Exception_Then_Operation_Throw_Excepti
...
+            underTest.RemoveFilter(id);
+
+            //then
```
Comment (vyshkouski-godeltech-intelliflo) [inline GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL.Tests/Services/FilterServiceTests.cs]: @IvanShetkoGodel Use assert please

File: GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL.Tests/Services/FilterServiceTests.cs [hunk 13]
```diff
@@ -190,5 +190,73 @@ public void Create_When_Filter_Added_With_Exception_Then_Operation_Throw_Excepti
...
+        [Test]
+        public void RemoveFilter_When_Filter_Exist_Should_Call_Delete_And_Save_Methods()
+        {
+            //arrange
+            int id = 1;
+            Filter toRemove = new Filter() { Id = 1};
+
+            unitOfWork.Setup(x => x.Filters.GetById(id)).Returns(toRemove);
+
+            //when
```
Comment (vyshkouski-godeltech-intelliflo) [inline GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL.Tests/Services/FilterServiceTests.cs]: @IvanShetkoGodel The same

File: GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL.Tests/Services/FilterServiceTests.cs [hunk 14]
```diff
@@ -190,5 +190,73 @@ public void Create_When_Filter_Added_With_Exception_Then_Operation_Throw_Excepti
...
+            underTest.RemoveFilter(id);
+
+            //then
```
Comment (vyshkouski-godeltech-intelliflo) [inline GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL.Tests/Services/FilterServiceTests.cs]: @IvanShetkoGodel The same

File: GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL.Tests/Services/FilterServiceTests.cs [hunk 15]
```diff
@@ -190,5 +190,73 @@ public void Create_When_Filter_Added_With_Exception_Then_Operation_Throw_Excepti
...
+        public void RemoveFilter_When_Exception_Appears_In_Delete_Method_Should_Call_Rollback_Method()
+        {
+            //arrange
+            int id = 1;
+            Filter toRemove = new Filter() { Id = 1 };
+
+            unitOfWork.Setup(x => x.Filters.GetById(id)).Returns(toRemove);
+            unitOfWork.Setup(x => x.Filters.Delete(toRemove)).Throws(new Exception());
+
+            //when
```
Comment (vyshkouski-godeltech-intelliflo) [inline GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL.Tests/Services/FilterServiceTests.cs]: @IvanShetkoGodel The same

File: GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL.Tests/Services/FilterServiceTests.cs [hunk 16]
```diff
@@ -190,5 +190,73 @@ public void Create_When_Filter_Added_With_Exception_Then_Operation_Throw_Excepti
...
+            underTest.RemoveFilter(id);
+
+            //then
```
Comment (vyshkouski-godeltech-intelliflo) [inline GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL.Tests/Services/FilterServiceTests.cs]: @IvanShetkoGodel The same

File: GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL.Tests/Services/FilterServiceTests.cs [hunk 17]
```diff
@@ -190,5 +190,73 @@ public void Create_When_Filter_Added_With_Exception_Then_Operation_Throw_Excepti
...
+        {
+            //arrange
+            int id = 1;
+            Filter toRemove = new Filter() { Id = 1 };
+
+            unitOfWork.Setup(x => x.Filters.GetById(id)).Returns(toRemove);
+            unitOfWork.Setup(x => x.Filters.Delete(toRemove));
+            unitOfWork.Setup(x => x.SaveChanges()).Throws(new Exception());
+
+            //when
```
Comment (vyshkouski-godeltech-intelliflo) [inline GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL.Tests/Services/FilterServiceTests.cs]: @IvanShetkoGodel The same

File: GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL.Tests/Services/FilterServiceTests.cs [hunk 18]
```diff
@@ -190,5 +190,73 @@ public void Create_When_Filter_Added_With_Exception_Then_Operation_Throw_Excepti
...
+            underTest.RemoveFilter(id);
+
+            //then
```
Comment (vyshkouski-godeltech-intelliflo) [inline GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL.Tests/Services/FilterServiceTests.cs]: @IvanShetkoGodel The same

File: GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL/Services/AuthenticationService.cs [hunk 19]
```diff
@@ -96,12 +96,11 @@ public AuthenticationService(IUnitOfWork unitOfWork, IEmailProvider emailProvide
         
         public async Task<ClaimsIdentity> Authenticate(UserDto userDto)
         {
-            logger.Info($"Cheking credentials for user {userDto.Email}");
-
             if (userDto == null)
             {
                 throw new ArgumentNullException(nameof(userDto));
             }
+            logger.Info($"Cheking credentials for user {userDto.Email}");
```
Comment (vyshkouski-godeltech-intelliflo) [inline GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL/Services/AuthenticationService.cs]: @IvanShetkoGodel Grammar mistake 'checking'

File: GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL/Services/FilterService.cs [hunk 20]
```diff
@@ -68,5 +68,25 @@ public FilterDto GetFilterById(int filterId)
                 throw new DataException(e.Message);
             }
         }
+        public async Task RemoveFilter(int id)
+        {
+            var filter = unitOfWork.Filters.GetById(id);
+            if (filter == null)
+            {
+                throw new ArgumentNullException(nameof(filter));
```
Comment (vyshkouski-godeltech-intelliflo) [inline GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL/Services/FilterService.cs]: @IvanShetkoGodel Invalid exception usage. Ask @vladkuzmich for more details

File: GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL/Services/FilterService.cs [hunk 21]
```diff
@@ -68,5 +68,25 @@ public FilterDto GetFilterById(int filterId)
...
+            }
+
+            logger.Info($"Remove filter id: {id}");
```
Comment (vyshkouski-godeltech-intelliflo) [inline GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL/Services/FilterService.cs]: @IvanShetkoGodel Invalid sense. Maybe Remove filter with id: {id}?

File: GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL/Services/FilterService.cs [hunk 22]
```diff
@@ -68,5 +68,25 @@ public FilterDto GetFilterById(int filterId)
...
+            try
+            {
+                unitOfWork.Filters.Delete(filter);
+                await unitOfWork.SaveChanges();
+            }
+            catch(Exception e)
+            {
+                logger.Error($"Remove filter {id} exception: {e.Message}");
```
Comment (vyshkouski-godeltech-intelliflo) [inline GodelMastery.FleaMarket/GodelMastery.FleaMarket.BL/Services/FilterService.cs]: @IvanShetkoGodel Change please message to make it more user friendly, something like:
Filter with {id} can't be removed due to exception: {e.Message}

File: GodelMastery.FleaMarket/GodelMastery.FleaMarket.Web/Views/Filter/RemoveFilter.cshtml [hunk 23]
```diff
@@ -0,0 +1,22 @@
+﻿@model GodelMastery.FleaMarket.Web.ViewModels.FilterViewModel
+
+<div>
+    @using (Html.BeginForm())
+    {
+        @Html.AntiForgeryToken()
+        <div class="modal-dialog" role="document">
+            <div class="modal-content">
+                <div class="modal-header">
+                    <h4 class="modal-title">Filter delete</h4>
```
Comment (vyshkouski-godeltech-intelliflo) [inline GodelMastery.FleaMarket/GodelMastery.FleaMarket.Web/Views/Filter/RemoveFilter.cshtml]: @IvanShetkoGodel Change please to 'Delete filter'

**Comment by IvanShetkoGodel** [comment]
@vyshkouski-godeltech-intelliflo Mentioned comments were fixed

**Pull request merged**